"""
@version:0.1
@author: Jetse

The Samtools module contains all rules for manipulating bam files with samtools.

A configuration file like the following is expected in a global CONFIG variable:
{
    "mapping":{
        "referenceGenome":"/path/to/referenceGenome.fasta
    }
}
"""
import subprocess, re
SAMTOOLS_VERSION = re.compile("Version:(.*?)\n").search(subprocess.Popen(CONFIG["samtools"]["path"] + 'samtools', shell=True, stderr=subprocess.PIPE).stderr.read().decode("utf-8")).group(1).strip()

# Download and install samtools
rule samtools_installation:
    output: CONFIG["samtools"]["path"] + 'samtools'
    shell:
        "TEMPDIR=`mktemp -d`;"
        "cd $TEMPDIR;"
        "git clone "+CONFIG['samtools']['URL']+" && "
        "git clone "+CONFIG['htslib']['URL']+" && "
        "cd samtools && "
        "make && "
        "mv samtools {output} &&"
        "touch {output}/samtools && "
        "cd / && rm -rf $TEMPDIR;"

rule samToBam:
    input:
        SAMTOOLS=rules.samtools_installation.output,
        BAM="mapped/{sample}.sam"
    output: "mapped/{sample}.bam"
    version: SAMTOOLS_VERSION
    shell: "{input.SAMTOOLS} view -bS {input.BAM} > {output}"

rule indexBam:
    input:
        SAMTOOLS=rules.samtools_installation.output,
        BAM="mapped/sorted.{sample}.bam"
    output: "mapped/{sample}.bam.idx"
    version: SAMTOOLS_VERSION
    shell: "{input.SAMTOOLS} index {input.BAM}"

rule sortBam:
    input:
        SAMTOOLS=rules.samtools_installation.output,
        BAM="mapped/{sample}.bam"
    output: "mapped/sorted.{sample}.bam"
    version: SAMTOOLS_VERSION
    shell: "{input.SAMTOOLS} sort -o {input.BAM} {output} > {output}"

rule mergeBam:
    input:
        SAMTOOLS=rules.samtools_installation.output,
        BAM='expand(",".join("/processedbam/{sample}.bam", sample=SAMPLES)'
    output: "processedbam/allMerged.bam"
    version: SAMTOOLS_VERSION
    shell: "{input.SAMTOOLS} merge {output} {input.BAM}"

rule addMdTag:
    input:
        SAMTOOLS=rules.samtools_installation.output,
        BAM="mapped/{sample}.bam",
        REFERENCE=CONFIG["refGenome"]
    output: "mapped/md.{sample}.bam"
    version: SAMTOOLS_VERSION
    shell: "{input.SAMTOOLS} calmd {input.BAM} {input.REFERENCE} > {output}"

rule faidX:
    input:
        SAMTOOLS=rules.samtools_installation.output,
        FASTA="{prefix}.fasta"
    output: "{prefix}.fasta.fai"
    version: SAMTOOLS_VERSION
    shell: "{input.SAMTOOLS} faidx {input.FASTA}"
